src/babel.js

```diff
@@ -20,248 +20,279 @@ import {
-              sourceFileName:
-                state.file.opts.sourceFileName || state.file.sourceFileName,
-              sourceMaps,
-              filename: state.file.opts.filename || state.file.filename
-            },
-            staticClassName: state.staticClassName,
-            isGlobal,
-            plugins: state.plugins,
-            vendorPrefixes
-          }
-          const splitRules =
-            typeof state.opts.optimizeForSpeed === 'boolean'
-              ? state.opts.optimizeForSpeed
-              : process.env.NODE_ENV === 'production'
+          path.replaceWith(makeStyledJsxTag(id, css))
+          return
+        }
+
+        const { vendorPrefixes, sourceMaps } = state.opts
+        const stylesInfo = {
+          ...state.styles.shift(),
+          fileInfo: {
+            file: state.file,
+            sourceFileName:
+              state.file.opts.sourceFileName || state.file.sourceFileName,
+            sourceMaps,
+            filename: state.file.opts.filename || state.file.filename
+          },
+          staticClassName: state.staticClassName,
+          isGlobal,
+          plugins: state.plugins,
+          vendorPrefixes
+        }
+        const splitRules =
+          typeof state.opts.optimizeForSpeed === 'boolean'
+            ? state.opts.optimizeForSpeed
+            : process.env.NODE_ENV === 'production'
 
-          const { hash, css, expressions } = processCss(stylesInfo, {
-            splitRules
-          })
+        const { hash, css, expressions } = processCss(stylesInfo, {
+          splitRules
+        })
 
-          path.replaceWith(makeStyledJsxTag(hash, css, expressions))
+        path.replaceWith(makeStyledJsxTag(hash, css, expressions))
+      }
+    }
+  }
+
+  // only apply JSXFragment visitor if supported
+  if (t.isJSXFragment) {
+    jsxVisitors.JSXFragment = jsxVisitors.JSXElement
+    jsxVisitors.JSXOpeningFragment = {
+      enter(path, state) {
+        if (!state.hasJSXStyle) {
+          return
         }
-      },
+
+        if (state.ignoreClosing === null) {
+          // We keep a counter of elements inside so that we
+          // can keep track of when we exit the parent to reset state
+          // note: if we wished to add an option to turn off
+          // selectors to reach parent elements, it would suffice to
+          // set this to `1` and do an early return instead
+          state.ignoreClosing = 0
+        }
+
+        state.ignoreClosing++
+      }
+    }
+  }
+
+  const visitors = {
+    inherits: jsx,
+    visitor: {
       Program: {
         enter(path, state) {
           state.hasJSXStyle = null
           state.ignoreClosing = null
           state.file.hasJSXStyle = false
 
           setStateOptions(state)
+
+          // we need to beat the arrow function transform and
+          // possibly others so we traverse from here or else
+          // dynamic values in classNames could be incorrect
+          path.traverse(jsxVisitors, state)
```

Comment:
the signature of `traverse` seems different:

```
export default function traverse(
  parent: Object | Array<Object>,
  opts?: Object,
  scope?: Object,
  state: Object,
  parentPath: Object,
) { }
```
https://github.com/babel/babel/blob/aaec2cd51dd24a1ecf52b82e0bbad0a5ccd60b6f/packages/babel-traverse/src/index.js#L14-L20

Reply:
oh probably `path.traverse`'s is different?

Reply:
Yeah `path.traverse` is a little different it passes along the extra params for you.

https://github.com/babel/babel/blob/a27b9b42990d44c4c2b5bf9e1184a10e49686a0d/packages/babel-traverse/src/path/index.js#L127